storage/src/main/java/tech/pegasys/teku/storage/store/BlockTree.java

```diff
@@ -0,0 +1,146 @@
+  private final HashTree hashTree;
+  final Map<Bytes32, UInt64> blockRootToSlot;
+
+  private BlockTree(final HashTree hashTree, final Map<Bytes32, UInt64> blockRootToSlot) {
+    validate(hashTree, blockRootToSlot);
+    this.hashTree = hashTree;
+    this.blockRootToSlot = blockRootToSlot;
+  }
+
+  public static BlockTree create(
+      final HashTree hashTree, final Map<Bytes32, UInt64> blockRootToSlot) {
+    return new BlockTree(hashTree, blockRootToSlot);
+  }
+
+  private void validate(final HashTree hashTree, final Map<Bytes32, UInt64> blockRootToSlot) {
+    checkArgument(
+        hashTree.size() == blockRootToSlot.size(),
+        "Slot lookup and hash tree must contain the same number of elements");
+    checkArgument(
+        Sets.difference(hashTree.getAllRoots(), blockRootToSlot.keySet()).isEmpty(),
+        "Slot lookup and hash tree must contain the same roots");
+  }
+
+  public BlockTree updated(final Bytes32 newRoot, Collection<SignedBeaconBlock> newBlocks) {
+    final HashTree updatedHashTree = hashTree.withRoot(newRoot).blocks(newBlocks).build();
+
+    // Create new root to slot mapping
+    final Map<Bytes32, UInt64> updatedBlockRootToSlot = new HashMap<>(blockRootToSlot);
+    newBlocks.forEach(b -> updatedBlockRootToSlot.put(b.getRoot(), b.getSlot()));
+    updatedBlockRootToSlot.keySet().removeIf(next -> !updatedHashTree.contains(next));
+
+    return new BlockTree(updatedHashTree, updatedBlockRootToSlot);
+  }
+
+  public HashTree getHashTree() {
+    return hashTree;
+  }
+
+  public Bytes32 getRootHash() {
+    return hashTree.getRootHash();
+  }
+
+  public boolean contains(final Bytes32 blockRoot) {
+    return hashTree.contains(blockRoot);
+  }
+
+  public Set<Bytes32> getAllRoots() {
+    return hashTree.getAllRoots();
+  }
+
+  /**
+   * @return A list of block roots ordered to guarantee that parent roots will be sorted earlier
+   *     than child roots
+   */
+  public List<Bytes32> getOrderedBlockRoots() {
+    return hashTree.breadthFirstStream().collect(Collectors.toList());
+  }
+
+  /**
+   * Returns true if the block root is at an internal epoch boundary. The rootHash is not considered
+   * an epoch boundary. A block is at an epoch boundary if it is the first block within an epoch.
+   *
+   * @param blockRoot The block root to check.
+   * @return True if the block root is at an internal epoch boundary.
+   */
+  public boolean isRootAtEpochBoundary(Bytes32 blockRoot) {
+    assertBlockIsInTree(blockRoot);
+    return hashTree
+        .getParent(blockRoot)
+        .map(
+            parentRoot -> {
+              final UInt64 blockEpoch = getEpoch(blockRoot);
+              final UInt64 parentEpoch =
+                  Optional.of(parentRoot)
+                      .filter(this::contains)
+                      .map(this::getEpoch)
+                      .orElse(blockEpoch);
+              return blockEpoch.isGreaterThan(parentEpoch);
+            })
+        .orElse(false);
```

Comment:
Isn't this:
```suggestion
    return hashTree
        .getParent(blockRoot)
        .filter(this::contains)
        .map(this::getEpoch)
        .map(parentEpoch -> getEpoch(blockRoot).isGreaterThan(parentEpoch))
        .orElse(false);
```

My reading of both versions being that it's the root at epoch boundary if this is the first block that exists for this epoch (so block's epoch is greater than the parent's epoch).

Reply:
yeah - your version is much simpler 👍